### ISSUE TITLE
MESOS: Possible to define Service port on Node

### ISSUE THREAD

**Issue by brandon-adams** [kubernetes/kubernetes#14911]
Sorry if the title is vague. I'm running Kubernetes v1 on Mesos (0.24), and I'm having some trouble getting service ports mapped correctly. Whenever I create a service to serve as an endpoint for a pod, it always chooses the first available node port on it's scheduled node. I have my range defined to start at port 1000, so it just chooses sequentially from there. Shouldn't it be using the value I provide? Or is this a limitation in the Mesos platform?

**Comment by bprashanth** [comment]
I'm confused by the explanantion, this should work if 30284 is available on every node in your cluster:

```yaml
apiVersion: v1
kind: Service
metadata:
  name: foo
  labels:
    app: foo
spec:
  type: NodePort
  ports:
  - port: 80
    nodePort: 30284
    targetPort: 8080
    protocol: TCP
    name: http
  selector:
    app: foo
```

**Comment by davidopp** [comment]
@kubernetes/kube-mesos

**Comment by davidopp** [comment]
https://github.com/kubernetes/kubernetes/blob/master/contrib/mesos/docs/issues.md
is probably relevant.

**Comment by karlkfi** [comment]
Yup, Service nodePort isn't supported on Mesos (yet). It will get assigned a (random-ish) port. Due to the way the Mesos Master filters offers and doesn't make them all available to all frameworks, it's possible that the k8s scheduler never gets offered all the ports in the range, or another framework uses it first.

We're still focusing on passing conformance tests. So this isn't a top priority yet. There's still some features we don't support in the rest of the e2e tests.

We'll make sure this is added to the known issues list and that there's a ticker for it in https://github.com/mesosphere/kubernetes-mesos, where we track our backlog.

**Comment by sttts** [comment]
I don't see that this is broken at all on Mesos. Set the service port range for the apiserver, e.g.
```
--service-node-port-range=30000-32767
```
and select a service nodePort, e.g.
```
nodePort: 30080
port: 80
```
The service will listen on the virtual IP on port 80 **and** on 30080 on the host interface. This is as expected.

**Comment by sttts** [comment]
@brandon-adams can you confirm the described behaviour for your setup?

**Comment by brandon-adams** [comment]
I think that it does work, though I might need some clarification as to the correct usage.

I've launched the apiserver with that flag, matching what I've defined as the available ports with Mesos. When I define my NodePort service, and it binds to the appropriate pod, it does indeed open up a port on the host according to it's definition. So if that's all it's meant to do, then we're straight.

However, I read somewhere (not sure at this point) that the service was meant to open the port across all nodes, and route traffic to the node on which the pod lives, and it does not do that. Perhaps I'm just confusing this behavior with something else.
